src/lib/staged_ledger/staged_ledger.ml

```diff
@@ -1983,11 +1983,96 @@ module T = struct
         : Ledger.unattached_mask ) ;
     r
 
+  module Application_state = struct
+    type txn =
+      ( Signed_command.With_valid_signature.t
+      , Zkapp_command.Valid.t )
+      User_command.t_
+
+    type t =
+      { valid_seq : txn Sequence.t
+      ; invalid : (txn * Error.t) list
+      ; skipped_by_fee_payer : txn list Account_id.Map.t
+      ; zkapp_space_remaining : int option
+      ; total_space_remaining : int
+      }
+
+    let init ?zkapp_limit ~total_limit =
+      { valid_seq = Sequence.empty
+      ; invalid = []
+      ; skipped_by_fee_payer = Account_id.Map.empty
+      ; zkapp_space_remaining = zkapp_limit
+      ; total_space_remaining = total_limit
+      }
+
+    let txn_key = function
+      | User_command.Zkapp_command cmd ->
+          Zkapp_command.(Valid.forget cmd |> fee_payer)
+      | User_command.Signed_command cmd ->
+          Signed_command.(forget_check cmd |> fee_payer)
+
+    let add_skipped_txn t (txn : txn) =
+      Account_id.Map.update t.skipped_by_fee_payer (txn_key txn)
+        ~f:(Option.value_map ~default:[ txn ] ~f:(List.cons txn))
+
+    let dependency_skipped txn t =
+      Account_id.Map.mem t.skipped_by_fee_payer (txn_key txn)
+
+    let try_applying_txn ?logger ~apply (state : t) (txn : txn) =
+      let open Continue_or_stop in
+      match (state.zkapp_space_remaining, txn) with
+      | _ when state.total_space_remaining < 1 ->
+          Stop (state.valid_seq, state.invalid)
+      | Some zkapp_limit, User_command.Zkapp_command _ when zkapp_limit > 1 ->
```

Comment:
It seems like when the match statement was refactored, this was messed up. This is the case where we are skipping, but it's predicated on the limit still being valid. I think the condition here is backwards.

Reply:
NIT: limit is a bit of a confusing name here, because this is space remaining, not the limit. It's just initialized to the limit, and is subtracted until it reaches 0.